### PR TITLE
(CDAP-17130) Added Distribution to BatchJoiner API

### PR DESCRIPTION
**Pull request by MEseifan**

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java [hunk 1]
```diff
@@ -302,7 +302,8 @@ public void testAutoInnerJoinSkewed() throws Exception {
                    .set("users_user_id", 1)
                    .set("users_name", "bob").build());
 
-    testSimpleAutoJoinSkewed(Arrays.asList("users", "interests"), expected);
+    testSimpleAutoJoinSkewed(Arrays.asList("users", "interests"), expected,  Engine.SPARK);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java]: nit: extra space before Engine

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java [hunk 2]
```diff
@@ -302,7 +302,8 @@ public void testAutoInnerJoinSkewed() throws Exception {
...
+    testSimpleAutoJoinSkewed(Arrays.asList("users", "interests"), expected,  Engine.MAPREDUCE);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java]: nit: extra space before Engine

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java [hunk 3]
```diff
@@ -36,8 +39,21 @@
    * @return returns join key
    * @throws Exception if there is some error getting the join key
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java:40]: include a `@deprecated` in the javadoc and mention that getJoinKeys() should be used instead.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java [hunk 4]
```diff
@@ -36,8 +39,21 @@
...
 
+  /**
+   * Return value for the join keys on which join will be performed
+   *
+   * @param stageName name of the stage to which records belongs to
+   * @param inputRecord input record to be joined
+   * @return returns a list of join keys
+   * @throws Exception if there is some error getting the join key
+   */
+  default Collection<JOIN_KEY> getJoinKeys(String stageName, INPUT_RECORD inputRecord) throws Exception{
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: add a space after Exception

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java [hunk 5]
```diff
@@ -309,13 +310,16 @@ protected PipeStage getSinkPipeStage(StageSpec stageSpec) throws Exception {
     public void transform(RecordInfo<INPUT_RECORD> input,
                           Emitter<KeyValue<OUT_KEY, TaggedWritable<OUT_VALUE>>> emitter) throws Exception {
       String stageName = input.getFromStage();
-      JOIN_KEY key = joiner.joinOn(stageName, input.getValue());
-      if (shouldFilter.test(stageName, key)) {
-        return;
+      Collection<JOIN_KEY> keys = joiner.getJoinKeys(stageName, input.getValue());
+      for (JOIN_KEY key : keys) {
+        if (shouldFilter.test(stageName, key)) {
+          return;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java]: return -> continue

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java [hunk 6]
```diff
@@ -36,8 +39,21 @@
    * @return returns join key
    * @throws Exception if there is some error getting the join key
    */
+  @Deprecated
   JOIN_KEY joinOn(String stageName, INPUT_RECORD inputRecord) throws Exception;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: can make this default as well, where the default implementation throws an exception. That way, plugins are not forced to implement this deprecated method.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java (hunk 1 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java]: done

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java (hunk 2 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/AutoJoinerTest.java]: done

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java (hunk 3 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java:40]: done

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java (hunk 4 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: done

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java (hunk 5 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java]: ah good catch, done

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java (hunk 6 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: done

**Comment by MEseifan** [comment]
https://builds.cask.co/browse/CDAP-DUT7292

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java [hunk 7]
```diff
@@ -17,26 +17,48 @@
 package io.cdap.cdap.etl.api;
 
 import io.cdap.cdap.api.annotation.Beta;
+import org.apache.commons.lang.NotImplementedException;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: avoid using classes from other libraries in the api modules. You can throw a UnsupportedOperationException instead.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java [hunk 8]
```diff
@@ -17,26 +17,48 @@
...
 
   /**
    * Return value for the join key on which join will be performed
    *
-   * @param stageName name of the stage to which records belongs to
+   * @param stageName   name of the stage to which records belongs to
    * @param inputRecord input record to be joined
    * @return returns join key
    * @throws Exception if there is some error getting the join key
+   * @deprecated getJoinKeys() should be used instead
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: `use {@link #getJoinKeys()} instead`

that way the javadoc will link to the method.

**Review by albertshau — APPROVED** [review]
minor comments, lgtm after they're fixed

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java (hunk 7 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: done

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java (hunk 8 above)
Comment (MEseifan) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/Joiner.java]: done

**Review by chtyim — APPROVED** [review]

**Pull request merged**